File: src/test/java/neoflix/_06_RatingMoviesTest.java
// Task: Rewrite the AuthService to allow users to authenticate against the database
// Outcome: A user will be able to authenticate against their database record
package neoflix;

import neoflix.services.RatingService;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.neo4j.driver.Driver;
import org.neo4j.driver.Values;

import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

class _06_RatingMoviesTest {
    private static Driver driver;

    private static String email = "graphacademy.reviewer@neo4j.com";
    private static String movieId = "769";
    private static String userId = "1185150b-9e81-46a2-a1d3-eb649544b9c4";
    private static int rating = 5;

    @BeforeAll
    static void initDriver() {
        AppUtils.loadProperties();
        driver = AppUtils.initDriver();

        driver.session().writeTransaction(tx -> tx.run("""
                MERGE (u:User {userId: $userId}) SET u.email = $email
                """, Values.parameters("userId", userId, "email", email)));
    }

    @AfterAll
    static void closeDriver() {
        driver.close();
    }

    @Test
    void writeMovieRatingAsInt() {
        RatingService ratingService = new RatingService(driver);

        var output = ratingService.add(userId, movieId, rating);

        assertNotNull(output);
        assertEquals(output.get("tmdbId"), movieId);
        assertEquals(Integer.parseInt(output.get("rating").toString()), rating);
    }
}

File: pkg/challenges/06_rating_movies_test.go
package challenges_test

import (
	"github.com/neo4j-graphacademy/neoflix/pkg/fixtures"
	"testing"

	"github.com/neo4j-graphacademy/neoflix/pkg/config"
	"github.com/neo4j-graphacademy/neoflix/pkg/services"
	"github.com/neo4j/neo4j-go-driver/v4/neo4j"
)

func TestRatingMovies(t *testing.T) {
	// Load Settings
	settings, err := config.ReadConfig("../../config.json")
	assertNilError(t, err)

	// Init Driver
	driver, err := config.NewDriver(settings)
	assertNilError(t, err)

	defer func() {
		assertNilError(t, driver.Close())
	}()

	// Create Services
	service := services.NewRatingService(
		&fixtures.FixtureLoader{Prefix: "../.."},
		driver)

	movieId := "769"
	userId := "1185150b-9e81-46a2-a1d3-eb649544b9c4"
	email := "graphacademy.reviewer@neo4j.com"
	rating := 5

	// Create the User
	session := driver.NewSession(neo4j.SessionConfig{})
	session.Run("MERGE (u:User {userId: $userId}) SET u.email = $email", map[string]interface{}{"userId": userId, "email": email})

	// Create the rating
	output, err := service.Save(rating, movieId, userId)

	assertNilError(t, err)
	assertEquals(t, movieId, output["tmdbId"])
	assertEquals(t, int64(rating), output["rating"])
}
